src/primitives/block.cpp

```diff
@@ -31,3 +31,12 @@ std::string CBlock::ToString() const
     }
     return s.str();
 }
+
+size_t GetVirtualBlockSize(const CBlock& block)
+{
+    // The formula is: vsize = base_size + witness_size / 4.
+    // We can only serialize base or totalbase+witness, however, so the formula
+    // becomes: vsize = base_size + (total_size - base_size) / 4 or
+    // vsize = (total_size + 3 * base_size) / 4.
+    return (::GetSerializeSize(block, SER_NETWORK, 0) * 3 + ::GetSerializeSize(block, SER_NETWORK, SERIALIZE_TRANSACTION_WITNESS) + 3) / 4;
```

Comment:
That's a round up; the formula should probably be clarified, or we should
switch to Nicolas's cost based constraints.